File: src/backend/vesicle.py
import math
import numpy as np
from .trackable import Trackable
from .constants import FARADAY_CONSTANT
from ..nestconf import Configurable
from typing import Optional


class Vesicle(Configurable, Trackable):
    # Configuration fields defined directly in the class
    specific_capacitance: float = 1e-2
    init_voltage: float = 4e-2
    init_radius: float = 1.3e-6
    init_pH: float = 7.4

    # Non-config fields
    TRACKABLE_FIELDS = ('pH', 'volume', 'area', 'capacitance', 'charge', 'voltage')

    def __init__(self,
                 *,               
                 display_name: str = None,
                 init_voltage: Optional[float] = None,
                 **kwargs):
        Configurable.__init__(self, **kwargs)
        Trackable.__init__(self, display_name=display_name)

        # Use typical values for voltage_exponent and half_act_voltage
        voltage_exponent = 80.0
        half_act_voltage = -0.04
        MAX_VOLTAGE = 709 / voltage_exponent + half_act_voltage

        # Clamp init_voltage if it exceeds the calculated safe limit
        if init_voltage is not None:
            if init_voltage > MAX_VOLTAGE:
                print(f"Warning: init_voltage {init_voltage} exceeds the safe limit. Clamping to {MAX_VOLTAGE}.")
                init_voltage = MAX_VOLTAGE
            elif init_voltage < -MAX_VOLTAGE:
                print(f"Warning: init_voltage {init_voltage} is below the negative safe limit. Clamping to {-MAX_VOLTAGE}.")
                init_voltage = -MAX_VOLTAGE
        else:
            init_voltage = self.init_voltage

        self.init_volume = (4 / 3) * math.pi * (self.init_radius ** 3)
        self.volume = self.init_volume

        self.init_area = 4.0 * math.pi * (self.init_radius ** 2)
        self.area = self.init_area

        self.init_capacitance = self.init_area * self.specific_capacitance
        self.capacitance = self.area * self.specific_capacitance

        self.init_charge = init_voltage * self.init_capacitance
        self.charge = init_voltage * self.capacitance

        self.pH = self.init_pH
        self.voltage = init_voltage

File: cpp_backend/src/Vesicle.cpp
#include "Vesicle.h"
#include <cmath>
#include <iostream>

// Constants for PI
const double PI = 3.14159265358979323846;

Vesicle::Vesicle(
    double initRadius,
    double initVoltage,
    double initPH,
    double specificCapacitance,
    const std::string& displayName
) : specificCapacitance_(specificCapacitance),
    initVoltage_(initVoltage),
    initRadius_(initRadius),
    initPH_(initPH),
    displayName_(displayName) {
    
    // Safety check for voltage - same as in Python implementation
    double voltageExponent = 80.0;
    double halfActVoltage = -0.04;
    double MAX_VOLTAGE = 709 / voltageExponent + halfActVoltage;
    
    if (initVoltage > MAX_VOLTAGE) {
        std::cout << "Warning: init_voltage " << initVoltage 
                  << " exceeds the safe limit. Clamping to " << MAX_VOLTAGE << "." << std::endl;
        initVoltage_ = MAX_VOLTAGE;
    } else if (initVoltage < -MAX_VOLTAGE) {
        std::cout << "Warning: init_voltage " << initVoltage 
                  << " is below the negative safe limit. Clamping to " << -MAX_VOLTAGE << "." << std::endl;
        initVoltage_ = -MAX_VOLTAGE;
    }
    
    // Calculate initial properties
    initVolume_ = (4.0 / 3.0) * PI * std::pow(initRadius_, 3);
    volume_ = initVolume_;
    
    initArea_ = 4.0 * PI * std::pow(initRadius_, 2);
    area_ = initArea_;
    
    initCapacitance_ = initArea_ * specificCapacitance_;
    capacitance_ = area_ * specificCapacitance_;
    
    initCharge_ = initVoltage_ * initCapacitance_;
    charge_ = initVoltage_ * capacitance_;
    
    pH_ = initPH_;
    voltage_ = initVoltage_;
}

void Vesicle::updateVolume(double newVolume) {
    volume_ = newVolume;
}

void Vesicle::updateArea() {
    // Calculate area from volume assuming sphere
    // A = (36π)^(1/3) * V^(2/3)
    area_ = std::pow(36.0 * PI, 1.0/3.0) * std::pow(volume_, 2.0/3.0);
}

void Vesicle::updateCapacitance() {
    capacitance_ = area_ * specificCapacitance_;
}

void Vesicle::updateCharge(double newCharge) {
    charge_ = newCharge;
}

void Vesicle::updateVoltage() {
    voltage_ = charge_ / capacitance_;
}

void Vesicle::updatePH(double newPH) {
    pH_ = newPH;
} 